storm/src/main/storm/mesos/resources/ResourceEntries.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package storm.mesos.resources;
+
+public class ResourceEntries {
+  public static final class ScalarResourceEntry implements ResourceEntry<Double> {
+    private ReservationType reservationType;
+    private Double value;
+
+    public ReservationType getReservationType() {
+      return reservationType;
+    }
+
+    public ScalarResourceEntry(Double value) {
+      this.value = value;
+    }
+
+    public ScalarResourceEntry(ReservationType reservationType, Double value) {
+      this.reservationType = reservationType;
+      this.value = value;
+    }
+
+    public Double getValue() {
+      return value;
+    }
+
+    @Override
+    public ScalarResourceEntry add(ResourceEntry<Double> resourceEntry) {
+      ScalarResourceEntry scalarResourceEntry = (ScalarResourceEntry) resourceEntry;
+      this.value += scalarResourceEntry.getValue();
+      return this;
+    }
+
+    @Override
+    public ScalarResourceEntry remove(ResourceEntry<Double> resourceEntry) {
+      ScalarResourceEntry scalarResourceEntry = (ScalarResourceEntry) resourceEntry;
+      this.value -= scalarResourceEntry.getValue();
+      return this;
+    }
+  }
+
+  public static final class RangeResourceEntry implements ResourceEntry<Long> {
+    private ReservationType reservationType;
+    private Long begin;
+    private Long end;
+
+    public RangeResourceEntry(Long begin, Long end) {
+      this.reservationType = reservationType;
+      this.begin = begin;
+      this.end = end;
+    }
+
+    public RangeResourceEntry(ReservationType reservationType, Long begin, Long end) {
+      this.reservationType = reservationType;
+      this.begin = begin;
+      this.end = end;
+    }
+
+    public Long getBegin() {
+      return begin;
+    }
+
+    public Long getEnd() {
+      return end;
+    }
+
+    public ReservationType getReservationType() {
+      return reservationType;
+    }
+
+    @Override
+    public RangeResourceEntry add(ResourceEntry<Long> resourceEntry) {
```

Comment:
This comment is not accurate either. I'm going to be a bit more clear here, so that it's less confusing with random variables (I chose poor names in my previous comment). For the sake of clarity, let's suppose we are starting with range `[s,e]`, and we are trying to add in range `[s', e']`. In your comment you say this will work if at least one of the following is true `s' < s`, `e' > e` or `s' < s && e' > e`.

Let's look at this example with concrete numbers `[s, e]` -> `[2000,2005]` and `[s',e']` -> `[13000, 30000]`.
In this case `s = 2000`, `e = 2005`, `s' = 13000`, and `e' = 30000`, so we fall into the case where `e' > e`, so using the code below our range is then going to set to `[13000, 2005]` even though this range is invalid.

Because of this, we really should only expand the range when `e + 1 = s'` or `s + 1 = e'`, or (as I missed before) when the ranges overlap (even though the Mesos offers won't really ever allow that).